lib/services/graph-progress.js

```diff
@@ -0,0 +1,232 @@
+// Copyright © 2017 Dell Inc. or its subsidiaries.  All Rights Reserved.
+
+'use strict';
+
+module.exports = graphProgressServiceFactory;
+
+graphProgressServiceFactory.$provide = 'Services.GraphProgress';
+graphProgressServiceFactory.$inject = [
+    '_',
+    'Protocol.Events',
+    'Services.Waterline',
+    'Logger',
+    'Constants',
+    'Promise',
+    'Assert',
+    'GraphProgress'
+];
+
+function graphProgressServiceFactory(
+    _,
+    eventsProtocol,
+    waterline,
+    Logger,
+    Constants,
+    Promise,
+    assert,
+    GraphProgress
+) {
+    var logger = Logger.initialize(graphProgressServiceFactory);
+
+    function GraphProgressService() {
+    }
+
+    GraphProgressService.prototype._publishGraphProgress = function(
+        graph,
+        graphState,
+        swallowError
+    ) {
+        assert.object(graph, 'graph');
+        assert.string(graph.name , 'graph.name');
+        assert.string(graphState, 'graphState');
+
+        var graphDescription = 'Graph "' + graph.name + '" ' + graphState;
+        var progress = GraphProgress.create(graph, graphDescription);
+
+        return eventsProtocol.publishProgressEvent(
+            graph.instanceId,
+            progress.getProgressEventData()
+        )
+        .catch(function(error) {
+            if(swallowError) {
+                logger.error('Error publishing graph ' + graphState + ' progress event', {
+                    graphId: graph.instanceId,
+                    _status: graph._status,
+                    error: error
+                });
+            } else {
+                throw error;
+            }
+        });
+    };
+
+    /**
+     * Publish a graph started progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} graph
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphStarted = function(graph, swallowError) {
+        return this._publishGraphProgress(graph, 'started', swallowError);
+    };
+
+    /**
+     * Publish a graph finished progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} graph
+     * @param {String} graphState
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphFinished = function(
+        graph,
+        graphState,
```

Comment:
Just a little suggestion, follow it or not is both OK for this PR:

I think there will be other people has similar surprise on the first sight of comparing `publishGraphFinished` with `publishGraphStarted`. Then I realized the graph start always only has one kind of state, but the graph may finish with different state, like `succeed`, `failed`. So this should be the reason why you add an extra argument `graphState` for `publishGraphFinished`.

To mitigate the surprise, you could (1) add some comment to explain the reason (2) Change the variable name to a better one, such as `graphState`==>`graphFinishedState`